csharp/src/ParallelSelectInternal.cs

```diff
@@ -0,0 +1,223 @@
+                                                                                         Func<TInput, Task<TOutput>>                func,
+                                                                                         int                                        parallelism,
+                                                                                         [EnumeratorCancellation] CancellationToken cancellationToken)
+  {
+    // CancellationTokenSource used to cancel all tasks inflight upon errors
+    var globalCts    = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+    var iterationCts = CancellationTokenSource.CreateLinkedTokenSource(globalCts.Token);
+
+    // Output
+    var channel = Channel.CreateUnbounded<Task<TOutput>>();
+
+    // Semaphore to limit the parallelism
+    var sem = new SemaphoreSlim(parallelism);
+
+    [SuppressMessage("ReSharper",
+                     "PossibleMultipleEnumeration")]
+    [SuppressMessage("ReSharper",
+                     "AccessToDisposedClosure")]
+    async Task Run()
+    {
+      try
+      {
+        var runningTasks = new List<Task>();
+
+        await foreach (var x in enumerable.WithCancellation(iterationCts.Token))
+        {
+          await sem.WaitAsync(iterationCts.Token)
+                   .ConfigureAwait(false);
+          var task = Task.Run(async () =>
+                              {
+                                TOutput res;
+                                try
+                                {
+                                  res = await func(x)
+                                          .ConfigureAwait(false);
+                                }
+                                catch
+                                {
+                                  iterationCts.Cancel();
+                                  throw;
+                                }
+
+                                sem.Release();
+                                return res;
+                              },
+                              globalCts.Token);
+
+          runningTasks.Add(task);
+
+          await channel.Writer.WriteAsync(task,
+                                          globalCts.Token)
+                       .ConfigureAwait(false);
+        }
+
+        await runningTasks.WhenAll()
+                          .ConfigureAwait(false);
+      }
+      finally
+      {
+        channel.Writer.Complete();
+      }
+    }
+
+    var run = Task.Run(Run,
+                       globalCts.Token);
+
+    try
+    {
+      await foreach (var res in channel.Reader.ToAsyncEnumerable(globalCts.Token))
+      {
+        yield return await res.ConfigureAwait(false);
+      }
+
+      await run.ConfigureAwait(false);
+    }
+    finally
+    {
+      // Ensure all running tasks are actually aborted
+      globalCts.Cancel();
+    }
+
+    sem.Dispose();
+    iterationCts.Dispose();
+    globalCts.Dispose();
```

Comment:
Why aren't they in the finally ? Or declared with `using var` ?

Reply:
If there is an error, there are some tasks in-flight that still need those.

Reply:
It should not change this behavior in the finally just above while ensuring that it is called

Reply:
Indeed, it does work. Changed.